codec-http2/src/main/java/io/netty/handler/codec/http2/HpackHuffmanDecoder.java

```diff
@@ -64,183 +53,4656 @@
      * @throws Http2Exception EOS Decoded
      */
     public AsciiString decode(ByteBuf buf, int length) throws Http2Exception {
-        processor.reset();
-        buf.forEachByte(buf.readerIndex(), length, processor);
-        buf.skipBytes(length);
-        return processor.end();
+        if (length == 0) {
+            return AsciiString.EMPTY_STRING;
+        }
+        byte[] dest = new byte[length * 8 / 5];
+        int k = 0;
+        int state = 0;
+        int flags = 0;
+        for (int i = 0; i < length; i++) {
+            byte input = buf.readByte();
```

Comment:
I am pretty sure you can make this even fast when using a `ByteBufProcessor` as you process byte by byte anyway. This way the range-check / ref-cnt check only needs to be done once

Reply:
I can also do the change on top of your change if you prefer.

Reply:
I'd rather you try, because my version didn't show any measurable speed up.   Here's what I tried, for posterity:

```java
    public AsciiString decode(ByteBuf buf, final int length) throws Http2Exception {
        if (length == 0) {
            return AsciiString.EMPTY_STRING;
        }

        final class HuffProcessor implements ByteProcessor {
            byte[] dest = new byte[length * 8 / 5];
            int k;
            int state;
            int flags;

            @Override
            public boolean process(byte input) throws Exception {
                int index = (state << 4) | ((input & 0xFF) >>> 4);
                int row = HUFFS[index];
                flags = row & 0x00FF00;
                if ((flags & HUFFMAN_FAIL_SHIFT) != 0) {
                    throw BAD_ENCODING;
                } else if ((flags & HUFFMAN_EMIT_SYMBOL_SHIFT) != 0) {
                    dest[k++] = (byte) (row & 0xFF);
                }
                state = row >> 16;

                index = (state << 4) | (input & 0x0F);
                row = HUFFS[index];
                flags = row & 0x00FF00;
                if ((flags & HUFFMAN_FAIL_SHIFT) != 0) {
                    throw BAD_ENCODING;
                } else if ((flags & HUFFMAN_EMIT_SYMBOL_SHIFT) != 0) {
                    dest[k++] = (byte) (row & 0xFF);
                }
                state = row >> 16;
                return true;
            }
        }
        HuffProcessor proc = new HuffProcessor();
        buf.forEachByte(buf.readerIndex(), length, proc);
        buf.readerIndex(buf.readerIndex() + length);

        if ((proc.flags & HUFFMAN_COMPLETE_SHIFT) != HUFFMAN_COMPLETE_SHIFT) {
            throw BAD_ENCODING;
        }
        return new AsciiString(proc.dest, 0, proc.k, false);
    }
```

Reply:
Will do

Reply:
@carl-mastrangelo did you compare when the calling method is nested 10 levels deep (as I think might be the case when used within the http2 impl)? :)

Reply:
When I ran it this morning, the `_byte` method costed like 3 or 4 levels.  Don't remember the exact amount.   It's unlikely we can guess how deep users' code will be though.   Were you thinking of manually inlining?

Reply:
No I just mean that the difference between your `readByte()` and `ByteProcessor` versions will likely only show up if this `decode()` method is nested deep enough in the stack. As run it _might_ be the case that `readByte()` is benefiting from an artificially shallow stack within the benchmark.

It actually probably doesn't need to be too deep since maybe 4-5 levels below `readByte()` need to be inlined for the access check elision, which leaves 4-5 from the top of the stack. Maybe running in the JMH harness comes in just within this number, but I would think when run in an http2 pipeline context it would usually (always?) be deeper than that. You could tweak the benchmark class to add like 3-4 methods of indirection.

Reply:
@carl-mastrangelo also the decoder wasn't threadsafe before so why not store/reuse the processor in an instance field (like the prior one)?

Reply:
Let's merge this one first and then I can do a followup.